pkg/devicetemplateconverter/vendor_profile_fetcher.go

```diff
@@ -0,0 +1,59 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicetemplateconverter
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// VendorIDProfileFetcher handles the validation and fetching of end device profile in the Device Repository.
+type VendorIDProfileFetcher struct {
+	Component        Component
+	DeviceRepository ttnpb.DeviceRepositoryClient
```

Comment:
There's quite some (duplicate) code below for trying to get the template from `DeviceRepository` field here and otherwise dial DR and get it from there. This seems only be necessary for testing. Also `Component` is quite a dependency here.

What you really need is an interface to get a template. You could have two implementations: one that gets a peer conn (and takes an interface to get the peer conn and whether insecure is allowed for creds), and another (only in testing) to get it from some mock.

Reply:
I agree, with this in mind I took the approach of changing this quite a bit again. Made it to be its own package inside the DTC. Wrote the interface for just the `GetTemplate method` and basically removed the duplicate code from the fetchers' implementations